src/inet/UDPEndPointImplNetworkFramework.mm

```diff
@@ -0,0 +1,655 @@
+                break;
+
+            case nw_listener_state_cancelled:
+                if (err == CHIP_NO_ERROR) {
+                    err = CHIP_ERROR_CONNECTION_ABORTED;
+                }
+
+                dispatch_semaphore_signal(mListenerSemaphore);
+                break;
+            }
+        });
+
+        nw_listener_start(listener);
+        dispatch_semaphore_wait(mListenerSemaphore, DISPATCH_TIME_FOREVER);
+
+        if (CHIP_NO_ERROR == err) {
+            mListener = listener;
+        }
+
+        return err;
+    }
+
+    CHIP_ERROR UDPEndPointImplNetworkFramework::StartConnection(nw_connection_t aConnection)
+    {
+        __block CHIP_ERROR err = CHIP_NO_ERROR;
+
+        nw_connection_set_queue(aConnection, mConnectionQueue);
+
+        nw_connection_set_state_changed_handler(aConnection, ^(nw_connection_state_t state, nw_error_t error) {
+            DebugPrintConnectionState(state);
+
+            switch (state) {
+            case nw_connection_state_invalid:
+                err = CHIP_ERROR_INCORRECT_STATE;
+                nw_connection_cancel(aConnection);
+                break;
+
+            case nw_connection_state_preparing:
+                err = CHIP_ERROR_INCORRECT_STATE;
+                break;
+
+            case nw_connection_state_waiting:
+                nw_connection_cancel(aConnection);
+                break;
+
+            case nw_connection_state_failed:
+                err = CHIP_ERROR_POSIX(nw_error_get_error_code(error));
+                break;
+
+            case nw_connection_state_ready:
+                err = CHIP_NO_ERROR;
+                dispatch_semaphore_signal(mConnectionSemaphore);
+                break;
+
+            case nw_connection_state_cancelled:
+                if (err == CHIP_NO_ERROR) {
+                    err = CHIP_ERROR_CONNECTION_ABORTED;
+                }
+
+                dispatch_semaphore_signal(mConnectionSemaphore);
+                break;
+            }
+        });
+
+        nw_connection_start(aConnection);
+        dispatch_semaphore_wait(mConnectionSemaphore, DISPATCH_TIME_FOREVER);
+
+        if (CHIP_NO_ERROR == err) {
+            DebugPrintConnection(aConnection);
+
+            mConnection = aConnection;
+            HandleDataReceived(mConnection);
+        }
+        return err;
+    }
+
+    void UDPEndPointImplNetworkFramework::HandleDataReceived(const nw_connection_t & aConnection)
+    {
+        nw_connection_receive_completion_t handler = ^(dispatch_data_t content, nw_content_context_t context, bool is_complete, nw_error_t receive_error) {
+            dispatch_block_t schedule_next_receive = ^{
+                if (receive_error == nullptr) {
+                    HandleDataReceived(aConnection);
+                } else if (OnReceiveError != nullptr) {
+                    __auto_type error_domain = nw_error_get_error_domain(receive_error);
+                    errno = nw_error_get_error_code(receive_error);
```

Comment:
Pre-existing, but writing to errno like this seems weird...  Shouldn't we use a local var?  Followup for this.

Reply:
I can't recall the exact reason for this approach, but there was likely a specific requirement at the time.
Further investigation is needed to determine if it is still necessary. (followup)

Reply:
https://github.com/project-chip/connectedhomeip/issues/36495